Created dotnet2.0 Builder Image and Added /v2/specialized Endpoint to dotnet2.0 Envrionment

paraspatidar opened the pull request:
Created dotnet2.0 Builder Image and Added /v2/specialized Endpoint to dotnet2.0 Envrionment .

with following features 
1. Dotnetcore 2.0 Envrionment Builder :
       (1) **builder  and precompilation** during package creation via builder.(see readme.md for builder)
       (2) **Nuget Package support** on runtime without modifying environment, which can be consumed by fission function.(see readme.md for builder)

2. Dotnetcore 2.0 Envrionment Builder :  
     (1)**additional appsetting/config json** file read and use via fission function (see readme.md for envrionment)
     (2)  **Namespace support** along with backward compatibility(see readme.md for envrionment) 


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/fission/fission/1001)
<!-- Reviewable:end -->

life1347 commented:
@paraspatidar Thanks for your contribution!

I'm not very familiar with dotnet, I'll do my best to review it and may take some time.
However, I notice some minor issues, maybe you can address them while I review the PR.

1. Currently, the release build will fail if the builder directory doesn't put in the right place. 
Please move the second level builder directory (with lower case "builder" name) to the top level like other languages. For example,
```
(X)
environments/dotnet20/Builder/Builder/Dockerfile
(O)
environments/dotnet20/builder/Dockerfile
```

2. It would be great for providing an example and README for how to use dotnet20 buider to build binary. For example, java env provides README contains steps to create/build a package.

paraspatidar commented:
Updated Builder folder as per PR.
readme.md available at : fission/environments/dotnet20/builder/readme.md with logic and how to example.

life1347 commented:
Hi @paraspatidar 

Curretnyl the build server doesn't have C# environment, we use [multi-stages build](https://github.com/fission/fission/commit/44be0f6859175468506aca67a80fa53f011a29cf) feature to build dotnot image. 

And when I'm trying to build container with command 
```
docker build -t dotnet20-env
```
locally, I keep encountering following error. Would you mind taking a look?  Thanks

```
  Restore completed in 9.3 sec for /proj/fission-dotnet20.csproj.
Microsoft (R) Build Engine version 15.3.409.57025 for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

Builder.cs(2,7): error CS0246: The type or namespace name 'NugetWorker' could not be found (are you missing a using directive or an assembly reference?) [/proj/fission-dotnet20.csproj]
BuilderEngine/BuilderEngine.cs(3,7): error CS0246: The type or namespace name 'NugetWorker' could not be found (are you missing a using directive or an assembly reference?) [/proj/fission-dotnet20.csproj]
BuilderEngine/BuilderEngine.cs(9,7): error CS0246: The type or namespace name 'NuGet' could not be found (are you missing a using directive or an assembly reference?) [/proj/fission-dotnet20.csproj]
Model/FunctionSpecification.cs(1,7): error CS0246: The type or namespace name 'NugetWorker' could not be found (are you missing a using directive or an assembly reference?) [/proj/fission-dotnet20.csproj]
Utility/Logger.cs(1,7): error CS0246: The type or namespace name 'log4net' could not be found (are you missing a using directive or an assembly reference?) [/proj/fission-dotnet20.csproj]
Utility/Logger.cs(2,7): error CS0246: The type or namespace name 'NuGet' could not be found (are you missing a using directive or an assembly reference?) [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(12,18): error CS0101: The namespace 'Fission.DotNetCore.Api' already contains a definition for 'FissionContext' [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(59,18): error CS0101: The namespace 'Fission.DotNetCore.Api' already contains a definition for 'Logger' [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(101,18): error CS0101: The namespace 'Fission.DotNetCore.Api' already contains a definition for 'FissionHttpRequest' [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(92,17): error CS0101: The namespace 'Fission.DotNetCore.Api' already contains a definition for 'Severity' [/proj/fission-dotnet20.csproj]
Model/FunctionSpecification.cs(29,24): error CS0246: The type or namespace name 'DllInfo' could not be found (are you missing a using directive or an assembly reference?) [/proj/fission-dotnet20.csproj]
BuilderEngine/BuilderEngine.cs(25,14): error CS0246: The type or namespace name 'DllInfo' could not be found (are you missing a using directive or an assembly reference?) [/proj/fission-dotnet20.csproj]
Utility/Logger.cs(14,27): error CS0246: The type or namespace name 'NuGet' could not be found (are you missing a using directive or an assembly reference?) [/proj/fission-dotnet20.csproj]
Utility/Logger.cs(16,17): error CS0246: The type or namespace name 'ILog' could not be found (are you missing a using directive or an assembly reference?) [/proj/fission-dotnet20.csproj]
Utility/Logger.cs(42,25): error CS0246: The type or namespace name 'LogLevel' could not be found (are you missing a using directive or an assembly reference?) [/proj/fission-dotnet20.csproj]
Utility/Logger.cs(48,25): error CS0246: The type or namespace name 'ILogMessage' could not be found (are you missing a using directive or an assembly reference?) [/proj/fission-dotnet20.csproj]
Utility/Logger.cs(54,30): error CS0246: The type or namespace name 'LogLevel' could not be found (are you missing a using directive or an assembly reference?) [/proj/fission-dotnet20.csproj]
Utility/Logger.cs(61,30): error CS0246: The type or namespace name 'ILogMessage' could not be found (are you missing a using directive or an assembly reference?) [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(14,16): error CS0111: Type 'FissionContext' already defines a member called '.ctor' with the same parameter types [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(30,38): error CS0111: Type 'FissionContext' already defines a member called 'Build' with the same parameter types [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(39,18): error CS0111: Type 'FissionContext' already defines a member called 'GetSettings' with the same parameter types [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(49,24): error CS0111: Type 'FissionContext' already defines a member called 'GetSettingsJson' with the same parameter types [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(61,21): error CS0111: Type 'Logger' already defines a member called 'Write' with the same parameter types [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(66,21): error CS0111: Type 'Logger' already defines a member called 'WriteInfo' with the same parameter types [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(71,21): error CS0111: Type 'Logger' already defines a member called 'WriteWarning' with the same parameter types [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(76,21): error CS0111: Type 'Logger' already defines a member called 'WriteError' with the same parameter types [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(81,21): error CS0111: Type 'Logger' already defines a member called 'WriteCritical' with the same parameter types [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(86,21): error CS0111: Type 'Logger' already defines a member called 'WriteVerbose' with the same parameter types [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(104,18): error CS0111: Type 'FissionHttpRequest' already defines a member called '.ctor' with the same parameter types [/proj/fission-dotnet20.csproj]
Model/FissionContext.cs(112,23): error CS0111: Type 'FissionHttpRequest' already defines a member called 'BodyAsString' with the same parameter types [/proj/fission-dotnet20.csproj]
The command '/bin/sh -c cd /proj && ./project-build.sh' returned a non-zero code: 1
```

paraspatidar commented:
The issue here is that builder folder is inside the environment folder , and thus as per existing docker file when we do **COPY * /proj/**  , it copies both fission-environment and builder solution in same **/proj** folder.

thus , when we try to build the project from /**proj/fission-dotnet20.csproj** only , then it fails with all above error as all those mentioned classes are part of builder project and not fission project.

the possible solution are : 
1) when building environment image , remove builder folder out of it.
2)  build builder image separately.

Or other option is : 
1) Do not copy the builder folder solution in step **COPY * /proj/** 

Let me know which approach we would prefer so that i can modify and update PR.

paraspatidar commented:
@life1347 , let me know if there is anything required or issue during PR review.

life1347 commented:
@paraspatidar sorry, didn't reply to this PR soon. Now, I'm reviewing the PR and above are some issues I've found so far.

paraspatidar commented:
@life1347 , I will fix those typos and enable multi stage build to get builder.
Can you let me know , if you have used same sample func.cs file to build the package or some other file ?
can you  share that source package.zip somewhere ?

also FYI , during builder build : when i was building from original fission source , I found that many dependent packages which builder depends on , have been change and thus it breaks the build. and we have to manually fork old version of those packages and then do `go get`.

paraspatidar commented:
@life1347 
Checked in with following changes :
 
1. Builder docker Update to get builder executable from official fission builder.
2. Corrected readme.md example as it was causing build package failure due to compilation  , now packages should compile and build successfully.
3. modified builder to propagate compilation  error in package info itself.
4. Corrected readme.md formatting issues.

Pending : 
1. **do a compile in Builder** : We are _**already doing a compile in builder**_ to find any possible compilation issue during build phase itself , but if we want to reduce the cold start of function , then we would have to save the in memory compiled assembly from builder to some physical file in deployment package , so that fission env v2 endpoint can directly read the physical file and execute it , let me know if we are looking for those changes ?

I would recommend to current code a try and in future we can enhance in some separate PR  as per above mentioned suggestion.

paraspatidar commented:
@life1347  for that build status showing succeed even if the builder process got failed  , not sure if that would be handled in this repo itself ?
As if there is any error , in that builder.dll we are throwing exception , but somehow as per code in 'fission\builder\client\client.go' , it seems to take it as 200 and thus shows status as succeed.can we work together on this one ?

life1347 commented:
After some experiments, we can add this line `set -euxo pipefail` to dotnet20/builder/build.sh then builder will be able to handle the build error. ([reference](https://coderwall.com/p/fkfaqq/safer-bash-scripts-with-set-euxo-pipefail))
```
#!/bin/bash
set -euxo pipefail
....
```

paraspatidar commented:
@life1347 I have made requested changes and checked in the files 
(except for
`` `#!/bin/bash
set -euxo pipefail```

I have also made you collaborator on my repo so that you dont need to wait for any small changes if required.
please let me know if anything is required from myside.

paraspatidar commented:
@life1347 is there any action item required form myside , i see this PR is nto having any activity from a long time. Please let me know if you have any review comment for me ?

life1347 commented:
Hi, I manually add `set -euxo pipefail` to dotnet20/builder/build.sh but still cannot get it to work. Could you please to take a look at it?

The docker image tested with:
* runtime: life1347/fission-dotnet20:0.1
* runtime builder: life1347/fission-dotnet20-builder:0.1

Here is the zip file I tested with: [dotnet.zip](https://github.com/fission/fission/files/3048196/dotnet.zip)
Error log:
```
=== Build Logs ===command=build
env=[PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin HOSTNAME=dotnet-874256-745cc69dff-8269v DOTNET_874256_PORT=tcp://10.104.110.183:8000 DOTNET_874256_PORT_8000_TCP_ADDR=10.104.110.183 DOTNET_874256_PORT_8001_TCP_ADDR=10.104.110.183 KUBERNETES_SERVICE_PORT=443 KUBERNETES_PORT_443_TCP_PROTO=tcp KUBERNETES_PORT=tcp://10.96.0.1:443 DOTNET_874256_SERVICE_PORT_BUILDER_PORT=8001 DOTNET_874256_PORT_8000_TCP_PROTO=tcp DOTNET_874256_PORT_8000_TCP_PORT=8000 KUBERNETES_SERVICE_HOST=10.96.0.1 KUBERNETES_SERVICE_PORT_HTTPS=443 KUBERNETES_PORT_443_TCP_PORT=443 KUBERNETES_PORT_443_TCP_ADDR=10.96.0.1 DOTNET_874256_SERVICE_HOST=10.104.110.183 DOTNET_874256_SERVICE_PORT=8000 DOTNET_874256_SERVICE_PORT_FETCHER_PORT=8000 DOTNET_874256_PORT_8001_TCP=tcp://10.104.110.183:8001 DOTNET_874256_PORT_8001_TCP_PORT=8001 DOTNET_874256_PORT_8000_TCP=tcp://10.104.110.183:8000 DOTNET_874256_PORT_8001_TCP_PROTO=tcp KUBERNETES_PORT_443_TCP=tcp://10.96.0.1:443 ASPNETCORE_URLS=http://+:80 DOTNET_RUNNING_IN_CONTAINER=true ASPNETCORE_VERSION=2.2.3 HOME=/root SRC_PKG=/packages/dotnet-zip-f22i-ve98cu DEPLOY_PKG=/packages/dotnet-zip-f22i-ve98cu-x7rvlj]
Builder Task Begins!
going to create logger!!
detailed logs for this build will be at : logs/2019_04_05_039b03a4-3416-4e42-8975-abc6df1304e6.log!!
Latest Package form Exact Search : CsvHelper| 12.1.2 in Repo Nugetdownload
---------------------------------------------------------------------
Latest Package form Exact Search : System.Reflection.TypeExtensions| 4.4.0 in Repo Nugetdownload
---------------------------------------------------------------------
Latest Package form Exact Search : Microsoft.CSharp| 4.4.0 in Repo Nugetdownload
---------------------------------------------------------------------
Package System.Reflection.TypeExtensions is Already Installed at  | /app/Nugetdownload/System.Reflection.TypeExtensions.4.4.0  | skipping instalation !!
Package Microsoft.CSharp is Already Installed at  | /app/Nugetdownload/Microsoft.CSharp.4.4.0  | skipping instalation !!
/app/Nugetdownload/Microsoft.CSharp.4.4.0/Microsoft.CSharp.4.4.0.nupkg
/app/Nugetdownload/System.Reflection.TypeExtensions.4.4.0/System.Reflection.TypeExtensions.4.4.0.nupkg
done for package System.Reflection.TypeExtensions , with total Dlls 0
done for package Microsoft.CSharp , with total Dlls 0
Package CsvHelper is Already Installed at  | /app/Nugetdownload/CsvHelper.12.1.2  | skipping instalation !!
/app/Nugetdownload/CsvHelper.12.1.2/CsvHelper.12.1.2.nupkg
dll Path: /app/Nugetdownload/CsvHelper.12.1.2/lib/netstandard2.0/CsvHelper.dll
done for package CsvHelper , with total Dlls 1
Total Dlls 1 for rootpackage CsvHelper-
done with nuget engine!!!!
DLL Info Gathered!!
Trying to compile it during build itslef !!
Compile Failed!!!!
Compilation result Gathered as : False!!
Compilation failed , throwing exception !!
COMPILATION ERROR : CS0103: The name 'respo' does not exist in the current context
COMPILATION ERROR : CS0103: The name 'respo' does not exist in the current context
COMPILATION ERROR : CS0103: The name 'respo' does not exist in the current context
Exception During Build :One or more errors occurred. (COMPILATION FAILED !! , See builder logs for details , total Errors :  3) |    at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at Builder.Builder.Main(String[] args) in /app/Builder.cs:line 31 |  Exception : LEVEL 1:
One or more errors occurred. (COMPILATION FAILED !! , See builder logs for details , total Errors :  3)
LEVEL 2:
COMPILATION FAILED !! , See builder logs for details , total Errors :  3|| STACK :   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at Builder.Builder.Main(String[] args) in /app/Builder.cs:line 31
2019/04/05 13:48:35 Error building source package: Error waiting for cmd 'build': exit status 134
+ cd /packages/dotnet-zip-f22i-ve98cu
+ echo src : /packages/dotnet-zip-f22i-ve98cu ,dest: /packages/dotnet-zip-f22i-ve98cu-x7rvlj
+ cd /app
+ dotnet Builder.dll

Unhandled Exception: System.AggregateException: One or more errors occurred. (COMPILATION FAILED !! , See builder logs for details , total Errors :  3) ---> System.Exception: COMPILATION FAILED !! , See builder logs for details , total Errors :  3
   at Builder.Engine.BuilderEngine.BuildPackage() in /app/BuilderEngine/BuilderEngine.cs:line 68
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at Builder.Builder.Main(String[] args) in /app/Builder.cs:line 31
/usr/local/bin/build: line 16:   805 Aborted                 (core dumped) dotnet Builder.dll
Error waiting for cmd 'build': exit status 134
2019/04/05 13:48:35 elapsed time in build request = 2.855511723s
```

life1347 commented:
Once the build problem is solved and add `set -euxo pipefail` to build.sh, then I think we are good to ship this.

vishal-biyani commented:
Hi @paraspatidar Can you please take a look at comments from @life1347 and take this further. Would really love to get this in Fission!

paraspatidar commented:
@vishal-biyani  sure , let me go through over it this weekend, however , rest of the things are already done , we are only left with 'set -euxo pipefail' and typo error in example program in readme.

vishal-biyani commented:
Thanks a lot, @paraspatidar - looking forward to it and want to to make this part of Fission 1.3 if all goes well.

paraspatidar commented:
@vishal-biyani @life1347  updated all docs and commands as per review comment and tested with same sample example (example also updated as it was having typo)  it is working as expected and package build status is also failing/success based on input function compilation.

Please let me know if that works ?

vishal-biyani commented:
Thanks @paraspatidar! @life1347 Not sure if you did try building this last time, but would have time and if you have some setup already?

life1347 commented:
@paraspatidar Thanks for the update, I will rebuild and give it a try again to see if I missed something.

paraspatidar commented:
> @paraspatidar Thanks for the update, I will rebuild and give it a try again to see if I missed something.

Please use the latest sample example as per readme.md as previous(3 month back) example was having typo

life1347 commented:
The example in `environments/dotnet20/README.md` works, however, the example in `environments/dotnet20/builder/README.md` still doesn't work for me.
```
Builder Task Begins!
going to create logger!!
detailed logs for this build will be at : logs/2019_05_09_c20c17b7-185a-4b72-bea0-4a976d538b7e.log!!
DLL Info Gathered!!
Trying to compile it during build itslef !!
Compile Failed!!!!
Compilation result Gathered as : False!!
Compilation failed , throwing exception !!
COMPILATION ERROR : CS0103: The name 'CsvHelper' does not exist in the current context
Exception During Build :
 One or more errors occurred. (COMPILATION FAILED !! , See builder logs for details , total Errors :  1) |    at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at Builder.Builder.Main(String[] args) in /app/Builder.cs:line 31 |
  Exception : LEVEL 1:
One or more errors occurred. (COMPILATION FAILED !! , See builder logs for details , total Errors :  1)
LEVEL 2:
COMPILATION FAILED !! , See builder logs for details , total Errors :  1|| STACK :   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at Builder.Builder.Main(String[] args) in /app/Builder.cs:line 31
+ cd /packages/funccsv-zip-hoz5-3757dl
+ echo src : /packages/funccsv-zip-hoz5-3757dl ,dest: /packages/funccsv-zip-hoz5-3757dl-w7ya8x
+ cd /app
+ dotnet Builder.dll
dotnet-1594690-86dd4d7bd9-rvsb6 builder
Unhandled Exception: System.AggregateException: One or more errors occurred. (COMPILATION FAILED !! , See builder logs for details , total Errors :  1) ---> System.Exception: COMPILATION FAILED !! , See builder logs for details , total Errors :  1
   at Builder.Engine.BuilderEngine.BuildPackage() in /app/BuilderEngine/BuilderEngine.cs:line 68
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at Builder.Builder.Main(String[] args) in /app/Builder.cs:line 31
/usr/local/bin/build: line 14:   186 Aborted                 (core dumped) dotnet Builder.dll
error waiting for cmd "build": exit status 134
{"level":"error","ts":1557393754.051754,"logger":"builder","caller":"builder/builder.go:124","msg":"error building source package","error":"error waiting for cmd \"build\": exit status 134","errorVerbose":"exit status 134\nerror waiting for cmd \"build\"\ngithub.com/fission/fission/builder.(*Builder).build\n\t/go/src/github.com/fission/fission/builder/builder.go:214\ngithub.com/fission/fission/builder.(*Builder).Handler\n\t/go/src/github.com/fission/fission/builder/builder.go:121\ngithub.com/fission/fission/builder.(*Builder).Handler-fm\n\tsrc/github.com/fission/fission/builder/cmd/main.go:48\nnet/http.HandlerFunc.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:1964\nnet/http.(*ServeMux).ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2361\nnet/http.serverHandler.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2741\nnet/http.(*conn).serve\n\t/usr/local/go/src/net/http/server.go:1847\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1333","stacktrace":"github.com/fission/fission/builder.(*Builder).Handler\n\t/go/src/github.com/fission/fission/builder/builder.go:124\ngithub.com/fission/fission/builder.(*Builder).Handler-fm\n\tsrc/github.com/fission/fission/builder/cmd/main.go:48\nnet/http.HandlerFunc.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:1964\nnet/http.(*ServeMux).ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2361\nnet/http.serverHandler.ServeHTTP\n\t/usr/local/go/src/net/http/server.go:2741\nnet/http.(*conn).serve\n\t/usr/local/go/src/net/http/server.go:1847"}
{"level":"info","ts":1557393754.05184,"logger":"builder","caller":"builder/builder.go:91","msg":"build request complete","elapsed_time":2.24576144}
{"level":"info","ts":1557393755.2543337,"logger":"builder","caller":"builder/builder.go:110","msg":"builder received request","request":{"srcPkgFilename":"funccsv-zip-hoz5-3757dl","command":"build"}}
{"level":"info","ts":1557393755.254385,"logger":"builder","caller":"builder/builder.go:112","msg":"starting build"}
```

life1347 commented on environments/dotnet20/README.md:
````diff
@@ -304,3 +268,98 @@ $ curl -XPOST http://localhost:8888/specialize
+				 in same namespace , this is also backward compatible , however main execution class
+				 would always be *FissionFunction* and its method *Execute*
+
+```
````
Would you mind to fix the indent in README (also in builder README.md)?

paraspatidar replied:
sure i will change indent.

for compilation error , did we put **CsvHelper** in nuget.txt ?
as in previous redme.md that is not highlighted well , and eve i missed to put its refrence and got same error.


Please have a look at this text in readme :
------------------------------------------------------------
Here is my nuget.txt

CsvHelper
As we dont want to exclude any specific dll thus we shall leave exclude.txt as empty
------------------------------------------------------------------------

paraspatidar replied:
updated readme by highlight this little more, to improve readability .

life1347 replied:
Yes, I put both nuget.txt and func.cs into the zip file. Here is the one I tested with https://github.com/fission/fission/files/3048196/dotnet.zip

paraspatidar replied:
Hi ,
the package which you shared , there  func.cs file was not having value initilized for string variable **respo** (that was issue in readme which i updated 3 month back's , so taking sample example reference from readme.md , here is the example ):

```
using System;
using Fission.DotNetCore.Api;

public class FissionFunction
{
    public string Execute(FissionContext context){
		string respo = "initial";
	        try
            {
				context.Logger.WriteInfo("Staring..... ");
				respo=$" sample object by getting Enum of   CsvHelper nuget dll: { CsvHelper.Caches.NamedIndex.ToString()}";
            }
            catch(Exception ex)
            {
				context.Logger.WriteError(ex.Message);
                respo = ex.Message;
            }
		context.Logger.WriteInfo("Done!");
		return respo;
    }
}
```

and nuget.txt is correct ,
thus after updating that correct C# function example in your zip ,i could successfully build the package :
Please use this zip :https://github.com/paraspatidar/SampleProjetcs/raw/master/func.zip

---------------------
```
Name:        func-zip-omoc
Environment: dotnetcor1
Status:      succeeded
Build Logs:
Builder Task Begins!
going to create logger!!
detailed logs for this build will be at : logs/2019_05_10_dafae657-3498-48a7-a0bb-ff6e1fd73017.log!!
Latest Package form Exact Search : CsvHelper| 12.1.2 in Repo Nugetdownload
---------------------------------------------------------------------
Latest Package form Exact Search : System.Reflection.TypeExtensions| 4.4.0 in Repo Nugetdownload
---------------------------------------------------------------------
Latest Package form Exact Search : Microsoft.CSharp| 4.4.0 in Repo Nugetdownload
---------------------------------------------------------------------
Package System.Reflection.TypeExtensions is Already Installed at  | /app/Nugetdownload/System.Reflection.TypeExtensions.4.4.0  | skipping instalation !!
Package Microsoft.CSharp is Already Installed at  | /app/Nugetdownload/Microsoft.CSharp.4.4.0  | skipping instalation !!
/app/Nugetdownload/System.Reflection.TypeExtensions.4.4.0/System.Reflection.TypeExtensions.4.4.0.nupkg
/app/Nugetdownload/Microsoft.CSharp.4.4.0/Microsoft.CSharp.4.4.0.nupkg
done for package Microsoft.CSharp , with total Dlls 0
done for package System.Reflection.TypeExtensions , with total Dlls 0
Package CsvHelper is Already Installed at  | /app/Nugetdownload/CsvHelper.12.1.2  | skipping instalation !!
/app/Nugetdownload/CsvHelper.12.1.2/CsvHelper.12.1.2.nupkg
dll Path: /app/Nugetdownload/CsvHelper.12.1.2/lib/netstandard2.0/CsvHelper.dll
done for package CsvHelper , with total Dlls 1
Total Dlls 1 for rootpackage CsvHelper-
done with nuget engine!!!!
DLL Info Gathered!!
Trying to compile it during build itslef !!
Compile Success!!
Compilation result Gathered as : True!!
Copy to Source Done!!
BuildSpecs Done!!
Builder Task Ends!
+ cd /packages/func-zip-omoc-uykmnm
+ echo src : /packages/func-zip-omoc-uykmnm ,dest: /packages/func-zip-omoc-uykmnm-llndku
+ cd /app
+ dotnet Builder.dll
+ cp -r /packages/func-zip-omoc-uykmnm /packages/func-zip-omoc-uykmnm-llndku
---------------------------------
 ```


Also , feel free to let me know for a session , if your would like to do this along side me,

life1347 replied:
After testing, found out that the problem that caused compilation error is that `nuget.txt` didn't allow any blank line in the file like following.
```
CsvHelper:12.1.2

```
which might be too sensitive for the user.

So we have some ways to address this problem when parsing `NugetPackageRegEx` & `ExcludeDllRegEx`:
1. Change regex in builderSetting.json
2. Read line by line and check whether it's a blank line

What do you think about it?

(btw, you may also want to fix indent of sample code in README. I pretty like [Code](https://code.visualstudio.com/) to edit markdown files.)

paraspatidar replied:
Thats the reason we mentioned about that regex to the user and also mentioned to leave a enter for last line.
As I remember i was facing some challenges last time for this newline thing in regex but I would try again to change that logic to read line by line to prevent user error.

In the meantime can you confirm about the functionality for builder ia as expected and can help fission 2.0 to have nugets on runtime and if we are good on that part?

life1347 replied:
Regex is good. But when it becomes complex, then it's hard for others to change it later. Also, as a tutorial doc, it's hard for every user to understand the regex in tutorial doc (though it's still helpful). Since the dependencies file is small, I think we can spend some time to parse them. Could [IsNullOrWhiteSpace](https://stackoverflow.com/a/6480081) gives any help in this case?

> In the meantime can you confirm about the functionality for builder ia as expected and can help fission 2.0 to have nugets on runtime and if we are good on that part?

Do you mean we keep the compilation at function cold start time? I'm ok to merge this PR first and creating an issue to track for dotnet builder enhancement. What do you think?

paraspatidar replied:
@life1347  , Made necessary changes for regex and blank line check .
Updated readme example with indent and new regex in builder readme.

now user doesnt need to bother about that new line and blank line for nugext and excludedll files.

Of course you can merge PR and we can then look forward for various option for cold start and compilation as there are many ways :
1. for example as builder has already tried compilation then we can improve environment cold start by saving that precompiled binary in deployment package and use it as it is in function thus saving recompilation at fission environment level.
2. many other possible option to improve cold start and compilation.


Lets see what users also face and suggest and we can track them as improvement item via github issues.

life1347 commented on environments/dotnet20/README.md:
```diff
@@ -304,3 +268,98 @@ $ curl -XPOST http://localhost:8888/specialize
+								var settings =context.GetSettings<SendGridSettings>("mysetting.json");
+								context.Logger.WriteInfo($"SendGridEndPoint port : {settings.SendGridEndPoints[0].port} ..... ");
+								respo=settings.SendGridEndPoints[0].port;           
+								context.Logger.WriteError("Done!!");
```
`WriteInfo`?

paraspatidar replied:
updated readme.md text , i have been using https://stackedit.io/app# , do u have any suggestion to use any better readme.md editor ?

life1347 replied:
I personally use [Code](https://code.visualstudio.com/) to edit markdown files.

life1347 commented:
LGTM! Would you mind to do rebase so that we can merge it? And, please file an issue for the discussion about compilation improvement.

paraspatidar commented:
@life1347  facing issue with rebase , i opted for merge `git merge upstream/master`  all evenrty is upto date with currunt master and aditional changes.
Can we merge the PR?

life1347 commented on environments/dotnet20/builder/nugetSettings.json:
```diff
@@ -0,0 +1,39 @@
+  "NugetFolder": "Nugetdownload",
+  "DisableCache": false,
+  "CSVDirectory": "logs",
+  //"NugetPackageRegEx": "\\:?\\s*(?<package>[^:\\n]*)(?:\\:)?(?<version>.*)?\\n",
```
Just notice this file and there are left comments. I'm checking why synk check failed, at the same time please remove unnecessary comments.

paraspatidar replied:
Nuget based  regex is being used as per regex in  builderSettings.json and not in nugetSettings.json (that nugetSettings.json regex is obsolete ) , removed it.
However , there is some other comments in this file which are kind of guidance for users on how to use private feed , i am keeping that comment for now so that user may understand it and add their private feeds for nuget.

life1347 replied:
ok, I originally thought it was sensitive info that pushed accidentally.

life1347 commented:
LGTM
